test/plugin/base_test.rb

```diff
@@ -1045,6 +1045,188 @@ def test_log_entry_trace_field
     end
   end
 
+  # Metadata Agent related tests.
+
+  # Test enable_metadata_agent not set or set to false.
+  def test_configure_enable_metadata_agent_default_and_false
+    setup_gce_metadata_stubs
+    [create_driver, create_driver(DISABLE_METADATA_AGENT_CONFIG)].each do |d|
+      assert_false d.instance.instance_variable_get(:@enable_metadata_agent)
+    end
+  end
+
+  # Test enable_metadata_agent set to true.
+  def test_configure_enable_metadata_agent_true
+    setup_gce_metadata_stubs
+    setup_metadata_agent_stubs
+    d = create_driver(ENABLE_METADATA_AGENT_CONFIG)
+    assert_true d.instance.instance_variable_get(:@enable_metadata_agent)
+    assert_requested_metadata_agent_stub(IMPLICIT_LOCAL_RESOURCE_ID)
+  end
+
+  # Test that an implicit monitored resource can be retrieved from Metadata
+  # Agent with an empty string as the local_resource_id.
+  def test_retrieve_implicit_monitored_resource
+    # GCE metadata stubs has VM_ID and ZONE, while the Metadata Agent stub has
+    # VM_ID2 and ZONE2.
+    setup_gce_metadata_stubs
+    setup_metadata_agent_stubs
+    setup_logging_stubs do
+      d = create_driver(ENABLE_METADATA_AGENT_CONFIG)
+      d.emit('message' => log_entry(0))
+      d.run
+    end
+    verify_log_entries(1, COMPUTE_PARAMS_WITH_DIFFERENT_VM_ID_AND_ZONE)
+    assert_requested_metadata_agent_stub(IMPLICIT_LOCAL_RESOURCE_ID)
+  end
+
+  # Docker Container.
+
+  # Test textPayload logs from Docker container stderr / stdout.
+  def test_docker_container_stderr_stdout_logs_text_payload
+    [1, 2, 3, 5, 11, 50].each do |n|
+      setup_gce_metadata_stubs
+      setup_metadata_agent_stubs
+      setup_logging_stubs do
+        d = create_driver(DOCKER_CONTAINER_CONFIG)
+        n.times do |i|
+          d.emit(docker_container_stdout_stderr_log_entry(log_entry(i)))
+        end
+        d.run
+      end
+      verify_log_entries(n, DOCKER_CONTAINER_PARAMS)
+      assert_requested_metadata_agent_stub(IMPLICIT_LOCAL_RESOURCE_ID)
+      assert_requested_metadata_agent_stub("container.#{DOCKER_CONTAINER_ID}")
+      WebMock.reset!
```

Comment:
I changed `setup_metadata_agent_stubs` to take a block. The naming or the method is left as it is, similar to `setup_logging_stubs`.

Reply:
Careful, `WebMock` covers all stubs, not just the metadata agent stubs. In fact, ideally, this would be part of the `setup_logging_stubs` context, because it also changes `@logs_sent`, but, unfortunately, we want to stop mocking the logging API early, to ensure `d.emit` is only called within that block... I wish Ruby allowed passing multiple blocks (one to execute the code and one to verify the results), but, alas, this seems to be [woefully inefficient](https://mudge.name/2011/01/26/passing-blocks-in-ruby-without-block.html)...
So, my suggestion was to create a top-level context object to wrap all stubs. Does that sound reasonable?

Reply:
Yes, `WebMock.reset!` will reset all stubs. I had to switch the order of a few lines to make sure that did not break out tests. If we are doing a top-level context object, we will be setting up **all** stubs for **all** tests that need any stub, right? That would make our test a bit cleaner. I wonder if it is a bit overkill though.

On the other hand, if we decide to do that, we should probably add `assert_expected_stub` checks for each case when a stub is used.

WDYT?

Reply:
Oh, sorry I wasn't clear. I was simply suggesting a stub context object that would initialize `@logs_sent` and reset `WebMock` at the end. We could still call `setup_*_stubs` in the block passed into the context to set up only the stubs we need. The `assert_expected_stub` calls you have now would also have to appear within the block (before the reset). See my earlier example... Does this make more sense?

Reply:
AH! I see what you meant now. Sorry I misunderstood. Changed in a82a36334b933299a81a058eaf91cddd32d0e2e7.